File: 594_Longest_Harmonious_Subsequence.py
from collections import Counter

class Solution(object):
    def findLHS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # Use Counter to get the frequency of each number
        counts = Counter(nums)

        max_length = 0
        # Iterate through the unique numbers and their counts
        for num_val, num_count in counts.items():
            # Check if num_val + 1 exists in the counts
            if (num_val + 1) in counts:
                # If it does, calculate the combined length and update max_length
                max_length = max(max_length, num_count + counts[num_val + 1])
        
        return max_length

File: 594_Longest_Harmonious_Subsequence.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int findLHS(int[] nums) {
        // Use a HashMap to store the frequency of each number
        Map<Integer, Integer> counts = new HashMap<>();
        for (int num : nums) {
            counts.put(num, counts.getOrDefault(num, 0) + 1);
        }

        int maxLength = 0;
        // Iterate through the unique numbers in the map
        for (int numVal : counts.keySet()) {
            // Check if numVal + 1 exists in the map.
            // If it does, we've found a potential harmonious pair.
            if (counts.containsKey(numVal + 1)) {
                // Calculate the combined length of the current number and its successor
                int currentLength = counts.get(numVal) + counts.get(numVal + 1);
                // Update maxLength if the current harmonious subsequence is longer
                maxLength = Math.max(maxLength, currentLength);
            }
        }
        return maxLength;
    }
}
